src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -566,32 +572,107 @@ private ValueTask<FlushResult> WriteDataAndTrailersAsync(Http2Stream stream, in
+            {
+                _outgoingFrame.HeadersFlags |= Http2HeadersFrameFlags.END_HEADERS;
+            }
 
-        while (!done)
-        {
-            _outgoingFrame.PrepareContinuation(Http2ContinuationFrameFlags.NONE, streamId);
+            WriteHeaderUnsynchronized();
+            _outputWriter.Write(dataToFrame[..currentSize]);
+            dataToFrame = dataToFrame.Slice(currentSize);
+        }
+    }
 
-            done = HPackHeaderWriter.ContinueEncodeHeaders(_hpackEncoder, _headersEnumerator, buffer, out payloadLength);
+    private void FinishWritingHeadersUnsynchronized(int streamId, int payloadLength, HeaderWriteResult writeResult)
+    {
+        Debug.Assert(payloadLength <= _maxFrameSize, "The initial payload lengths is written to _headerEncodingBuffer with size of _maxFrameSize");
+        byte[]? largeHeaderBuffer = null;
+        Span<byte> buffer;
+        if (writeResult == HeaderWriteResult.Done)
+        {
+            // Fast path, only a single HEADER frame.
             _outgoingFrame.PayloadLength = payloadLength;
-
-            if (done)
+            _outgoingFrame.HeadersFlags |= Http2HeadersFrameFlags.END_HEADERS;
+            WriteHeaderUnsynchronized();
+            _outputWriter.Write(_headerEncodingBuffer.AsSpan(0, payloadLength));
+            return;
+        }
+        else if (writeResult == HeaderWriteResult.MoreHeaders)
+        {
+            _outgoingFrame.PayloadLength = payloadLength;
+            WriteHeaderUnsynchronized();
+            _outputWriter.Write(_headerEncodingBuffer.AsSpan(0, payloadLength));
+        }
+        else
+        {
+            // This may happen in case of the TRAILERS after the initial encode operation.
+            // The _maxFrameSize sized _headerEncodingBuffer was too small.
+            while (writeResult == HeaderWriteResult.BufferTooSmall)
+            {
+                Debug.Assert(payloadLength == 0, "Payload written even though buffer is too small");
+                largeHeaderBuffer = ArrayPool<byte>.Shared.Rent(_headersEncodingLargeBufferSize);
+                buffer = largeHeaderBuffer.AsSpan(0, _headersEncodingLargeBufferSize);
+                writeResult = HPackHeaderWriter.RetryBeginEncodeHeaders(_hpackEncoder, _headersEnumerator, buffer, out payloadLength);
+                if (writeResult != HeaderWriteResult.BufferTooSmall)
+                {
+                    SplitHeaderFramesToOutput(streamId, buffer[..payloadLength], endOfHeaders: writeResult == HeaderWriteResult.Done, isFramePrepared: true);
+                }
+                else
+                {
+                    // The size is backed in a local field and not reset for further streams on the connections.
+                    // When a large header is repeatedly returned on a connection, the buffer size won't need to be repeatedly increased.
+                    _headersEncodingLargeBufferSize = checked(_headersEncodingLargeBufferSize * HeaderBufferSizeMultiplier);
```

Comment:
This will throw when the header is at least `maxint / 2` bytes?  What does it do when it throws?  I remember there was an earlier fix to make sure an error didn't result in an infinite loop of empty writes.  I'm pretty sure this won't cause that, but I thought I should confirm.

Reply:
Note that the header may be actually already larger to `int.MaxValue / 2`. For example a client updates the `_maxFrameSize` to a custom value and we end up here with `(2/3) * int.MaxValue` Of course the next increase would throw, but to be clear, it does not limit the max response header size.

When it throws there, this is what it would throw:

```
info: Microsoft.AspNetCore.Server.Kestrel.Http2[38]
      Connection id "0HN45U6KT28C4": HPACK encoding error while encoding headers for stream ID 1.
      System.OverflowException: Arithmetic operation resulted in an overflow.
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2FrameWriter.FinishWritingHeadersUnsynchronized(Int32 streamId, Int32 payloadLength, HeaderWriteResult writeResult) in D:\repos\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http2\Http2FrameWriter.cs:line 660
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2FrameWriter.WriteResponseHeadersUnsynchronized(Int32 streamId, Int32 statusCode, Http2HeadersFrameFlags headerFrameFlags, HttpResponseHeaders headers) in D:\repos\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http2\Http2FrameWriter.cs:line 523
```

which then ends up handled in `WriteResponseHeadersUnsynchronized` with a `ConnectionAbortedException`

But I could only get here by forcing a large initial value to `_headersEncodingLargeBufferSize` because in practise my machine throws an OOM way earlier, already when I try to allocate the string value of such large header.